### PR TITLE
📦 NEW: Add CoverageReporter for batching code coverage reports 

### PR DESCRIPTION
**Pull request by michaelborn**
## Problem: Large Codebases Cannot Generate Full Coverage Report

Due to memory limitations in CI environments, larger codebases are unable to run all tests as a single `testbox run` command. Instead, specs are run in a methodical folder-by-folder sequence, separating the testbox run out over many requests and thus working around the Out-Of-Memory exceptions.

While this works, it prevents accurate reporting of code coverage since only a small portion of the tests are executed during any request. The generated code coverage report only shows a tiny fraction of the coverage - say, 2% - and not the whole picture

## Solution: Aggregate + Combine Code Coverage

This PR introduces a `CoverageReporter` component which 

1. runs on every testbox code coverage execution
2. loads any previous coverage data from a JSON file
3. combines the previous coverage data with the current execution's coverage data (file by file and line by line)
5. persists the COMBINED coverage data to a JSON file.
6. returns the COMBINED coverage data for the `CoverageBrowser.cfc` to build as an HTML report

When setting `url.isBatched=true` and executing the batched test runner, the code coverage report will then grow with each sequential `testbox run` command.

### CODE REVIEW

**Comment by jclausen** [comment]
Go ahead and get the one test that's failing passing and I will review this tomorrow.  This is fantastic, though! Thank you for your hard work on this.

File: system/coverage/CoverageReporter.cfc [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+	/**
+	 * Begin processing coverage data IF batching enabled.
+	 *
+	 * @coverageQuery The result from CoverageGenerator's generateData() method.
+	 *
+	 * @return an aggregated, FULL report of all test coverage data accumulated to this point.
+	 */
+	public any function processCoverageReport( required any coverageQuery ){
+		if ( !getReportOptions().isBatched ) {
```
Comment (jclausen) [inline system/coverage/CoverageReporter.cfc:61]: I would just use a ternary conditional here with a single return statement

File: test-harness/runner.cfm [hunk 2]
```diff
@@ -17,6 +17,8 @@
 <cfparam name="url.coverageBlacklist"				default="/testbox,/coldbox,/tests,/modules,Application.cfc,/index.cfm">
 <!---<cfparam name="url.coverageBrowserOutputDir"		default="#expandPath( '/tests/results/coverageReport' )#">--->
 <!---<cfparam name="url.coverageSonarQubeXMLOutputPath"	default="#expandPath( '/tests/results/SonarQubeCoverage.xml' )#">--->
+<!--- Enable batched code coverage reporter, useful for large test bundles which require spreading over multiple testbox run commands. --->
+<!--- <cfparam name="url.isBatched"						default="false"> --->
```
Comment (jclausen) [inline test-harness/runner.cfm:21]: Remove commented code above

**Review by jclausen — CHANGES_REQUESTED** [review]
This looks really good!  A couple of cleanup comments.

File: test-harness/runner.cfm (hunk 2 above)
Comment (michaelborn) [reply test-harness/runner.cfm:21]: I hear you, but I feel like this needs to remain here for documentation reasons. Isn't this directory what `box testbox generate harness` uses to create a test harness?

Comment (bdw429s) [reply test-harness/runner.cfm:21]: I would prefer that stays-- those comments have served as documentation for the code coverage features people can opt into.  We'll need to ensure we have additions to the TestBox docs as well explaining how to use this feature.

Comment (jclausen) [reply test-harness/runner.cfm:21]: I'm fine with that.

Comment (michaelborn) [reply test-harness/runner.cfm:21]: > We'll need to ensure we have additions to the TestBox docs as well explaining how to use this feature.

Noted - some of the existing options are missing from the docs AND this file as well. I'm going to go ahead and add those.

**Comment by michaelborn** [comment]
@lmajano I fixed the failing ACF tests and merged latest just to be sure we wouldn't have conflicts.

(Oh, and, thanks for disabling the NULL tests!)

**Pull request merged**